### ISSUE TITLE
[Will Not Fix]  Hot end self test Error w/ stainless steel heatblock:"Selftest error! Please check: Heater/Thermistor Not connected"

### ISSUE THREAD

**Issue by hjschiller** [prusa3d/Prusa-Firmware#4139]
**Printer type** - MK3S+

** Problem description**
I have an MK3S+ with an upgraded E3D Thermistor, Upgraded E3D High Precision 40W Ceramic Heater Cartridges, and Stainless steel heatblock.

Self test fails when checking the hotend with the error "Selftest error! Please check: Heater/Thermistor Not connected.

Thermistor is connected and registers 19C at start of test, temp rises to about 30C, then displays "Selftest error! Please check: Heater/Thermistor Not connected".

All other calibration works fine including PID calibration at 240C, and thermal calibration.
Printer prints fine, but I get an annoying warning that the printer is not calibrated.

I suspect the issue is with the amount of time it takes to change the temperature of the heatblock.  The Stainless steel takes longer to transfer the heat than the stock aluminum heatblock.  Seems to take longer to get to the failure point with a  High Precision 30W Ceramic Heater Cartridge as. compared to High Precision 40W Ceramic Heater Cartridge .  I'm guessing this test uses an arbitrary pre-defined window, rather than the data from the PID calibration and/or thermal calibration.  

**Describe the solution you'd like**
Would like to be able to use upgraded components, such as a stainless steel heat block, and upgraded thermistor (at the time I replaced my heatblock an upgraded Prusa Thermistor was not available), and E3D High Precision 40W Ceramic Heater Cartridges.  

Ideally if PID and thermal calibration is successful, have a secret menu where you can either:
 1. override the Hotend self test or 
 2. clear the printer is not calibrated state or
 3. if it exists, use the PID/thermal calibration to define the selftest range or
 4. make the selftest range sufficiently wide to cover these non-standard config, 
     - possibly offer a soft error "This looks out of spec for prusa standard Heater/Thermistor/heatblock.  Confirm you are using standard parts y/n" or
  5. offer a secret setting to indicate you are using non-standard parts, and loosen the test requirements, or
  6. offer a prusa certified High Precision 40W Ceramic Heater Cartridge.

In short, as it is working just fine, please give me a way to clear/avoid this error condition.   
I have no objection to using Prusa parts if they have the same quality/performance.

**Comment by hjschiller** [comment]
Found my old beat up aluminum block, and ran the selftest with it, and a spare upgraded E3D Thermistor, Upgraded E3D High Precision 30W Ceramic Heater Cartridges dangling 
![IMG-8519](https://user-images.githubusercontent.com/77127997/231332355-5dc7ed34-56b4-4853-9f64-f15241cc0852.jpg)

Seems the Stainless steel heatblock was causing the failure.  

Would be nice to not have to disconnect and reconnect the thermistor and heater to get it to not have an error message.

**Comment by jasonyork** [comment]
I had a similar problem this week on my MK2.5.  My replacement heater block was causing the hotend portion of the selftest to fail because it wasn't heating fast enough.  Everything was working fine, but I couldn't clear the selftest message.  Here was my solution to bypass in case it helps anyone else:

1. Settings -> Temperature -> Nozzle -> 200 deg
2. Wait for nozzle to reach temp
3. Preheat -> Cooldown
4. Do not wait for nozzle to cooldown.  Immediately unplug the hotend thermistor
5. Calibration -> Selftest
6. The test should immediately start and not attempt to cooldown before starting
7. Right after the heatbed test passes, and it begins testing the hotend, at this point plug the thermistor back in.  This causes it to see a jump in temperature "tricking" it into thinking it heated up quickly.

I would love to see an option buried in the setting somewhere to just clear the "not calibrated" message.

**Comment by Prusa-Support** [comment]
Thanks for your suggestions.

> I have no objection to using Prusa parts if they have the same quality/performance.

@hjschiller genuine Prusa parts are surely reliable and suitable for everyday printing. These are quality parts.
Anyway, can you confirm that the thermal calibration won't fail and that no thermal anomaly will appear while printing?
Have you by any chance modified some parameters for the thermal model calibration to succeed?

Michele Moramarco
Prusa Research

**Comment by hjschiller** [comment]
@Prusa-Support 
Thermal calibration and PID calibration at 240C (and whatever the default is) just worked. No modification of anything. 

Have printed a bunch of things with the most recent code even an 8 hour print. All works fine. 

It just won’t pass the selftest despite running a proper calibration. There must be a hard coded time limit in the heatblock selftest to see the heat rise, and since the stainless steel takes 5x longer to heat up it bails out on the test when it doesn’t see a rise assuming either the thermistor or heat cartridge is not functioning. 

Beyond failing the selftest, it also complains that the self test failed. 

I got that to go away by hooking up my older stock aluminum heatblock and putting in an older heat cartridge and thermistor. 

I’m sure the stock parts are fine, but my aluminum block got very beaten up as you can see in the picture. The stainless heatblock takes a beating and cleans up nice. 

I think you now offer a Prusa high performance ceramic heat cartridge  (yeah!) and I did buy a thermistor from printed solid but it is currently in the spare drawer.

Some things that could work.
1. By pass the heatblock selftest after agreeing to the risks.  

2. Use the thermal calibration and/or pid calibration as a substitute for the passing the heatblock selftest. 

3. Use the data from the calibration (thermal and/or PID) to determine the expected heat up time for the heatblock. 

Failing the selftest prior to doing thermal calibration / PID calibration is perfectly acceptable. 

Happy to run any tests or collect more data if that helps. 

___Jason

**Comment by hjschiller** [comment]
@jasonyork can you confirm calibration and printing works fine for you as well, and no software modifications were made. 

___Jason

**Comment by jasonyork** [comment]
@hjschiller - Correct, after bypassing the calibration (using the method I outlined above) I've had no issues with my prints and I did not make any custom firmware modifications.

**Comment by Prusa-Support** [comment]
Thanks for clarifying.

Indeed the self-test checks the temperature against a timer.
The hotend is preheated at 40˚C at the beginning of the test. Once the target temperature is met, it continues heating at full power for 40 seconds. The expected temperature after 40s should be between 180˚C-230˚C for the nozzle heater to successfully complete the self-test.
This test is not so strict after all.

I hope this helps to responsibly walkaround the current limitations but please mind these limitations are set in place for safety reasons and we can't take responsibility for the use of third-party components.

Given the limited space in the Einsy memory, as of now, we don't expect to change this behavior so the users should compile a custom firmware with a custom target range at their own risk if they want to change this behavior.

I hope you can understand and consider closing the issue. It can be reopened in the future if it gains popularity.
For the time being, I'd leave it up to you to close the issue. 🙂 

Michele Moramarco
Prusa Research

**Comment by hjschiller** [comment]
@Prusa-Support , happy to close the issue. Is it possible to take this as a feature request for next generation hardware when and if memory constraints are less of an issue?  

Does this apply to the MK4 and XL?

**Comment by Prusa-Support** [comment]
Thanks for understanding.

I'm afraid it is too early to submit such a feature request on MK4/XL firmware, given that there is lack of compatible third-party components at the moment.
I suggest trying later if there will be a compatible third-party component that is worth using, if the problem will persist with it.
It is best to open issues based on individual real cases.

Michele Moramarco
Prusa Research

**Comment by hjschiller** [comment]
Updating subject from:
[ENHANCEMENT] Hidden Selftest failure override
to: 
  [Will Not Fix]  Hot end self test Error w/ stainless steel heatblock:"Selftest error! Please check: Heater/Thermistor Not connected"

In hopes that this issue will be more easily found if others experience this problem.

Closing the issue as per previous update, Rambo memory is not sufficient to add  self test error work around.

**Comment by jasonyork** [comment]
Also worth mentioning for anyone else that goes down this rabbit hole.. I'd recommend checking the power connectors on the mainboard from the power supply and to the heater.  They degrade over time and could contribute to performance problems.

**Comment by Prusa-Support** [comment]
Thank you all for your contribution and collaboration.

Michele Moramarco
Prusa Research
